eclair-core/src/main/scala/fr/acinq/eclair/payment/receive/MultiPartHandler.scala

```diff
@@ -298,20 +306,50 @@ object MultiPartHandler {
                   nodeParams.db.payments.addIncomingPayment(invoice, paymentPreimage, r.paymentType)
                   invoice
                 case r: ReceiveOfferPayment =>
-                  // TODO: get blinded paths from the router instead
-                  val pathId = RouteBlindingEncryptedDataTlv.PathId(randomBytes32())
-                  val dummyConstraints = RouteBlindingEncryptedDataTlv.PaymentConstraints(CltvExpiry(nodeParams.currentBlockHeight + 144), 1 msat)
-                  val dummyRelay = RouteBlindingEncryptedDataTlv.PaymentRelay(CltvExpiryDelta(0), 0, 0 msat)
-                  val dummyScid = RouteBlindingEncryptedDataTlv.OutgoingChannelId(ShortChannelId.toSelf)
-                  val dummyPath = Seq(
-                    (nodeParams.nodeId, RouteBlindingEncryptedDataCodecs.blindedRouteDataCodec.encode(TlvStream(dummyScid, dummyRelay, dummyConstraints)).require.bytes),
-                    (nodeParams.nodeId, RouteBlindingEncryptedDataCodecs.blindedRouteDataCodec.encode(TlvStream(dummyConstraints, pathId)).require.bytes),
-                  )
-                  val blindedRoute = Sphinx.RouteBlinding.create(randomKey(), dummyPath.map(_._1), dummyPath.map(_._2))
+                  val amount = r.invoiceRequest.amount.orElse(r.offer.amount.map(_ * r.invoiceRequest.quantity)).get
+                  val paths = r.routes.map(nodeIds => {
+                    require(nodeIds.nonEmpty, "route can't be empty")
+                    val pathId = randomBytes32()
+                    val finalExpiryDelta = nodeParams.channelConf.minFinalExpiryDelta + 3
+                    val finalConstraints = RouteBlindingEncryptedDataTlv.PaymentConstraints(finalExpiryDelta.toCltvExpiry(nodeParams.currentBlockHeight), nodeParams.channelConf.htlcMinimum)
+                    val zeroPaymentInfo = PaymentInfo(0 msat, 0, CltvExpiryDelta(0), nodeParams.channelConf.htlcMinimum, amount, Features.empty)
+                    val finalPayload = RouteBlindingEncryptedDataCodecs.blindedRouteDataCodec.encode(TlvStream(
+                      finalConstraints,
+                      RouteBlindingEncryptedDataTlv.PathId(pathId)
+                    )).require.bytes
+                    val (paymentInfo, payloads) = if (nodeIds.length > 1) {
+                      val timeout = 30 second
+                      val routeResponse = Await.result(r.router.ask(Router.FinalizeRoute(0 msat, Router.PredefinedNodeRoute(nodeIds)))(timeout).mapTo[Router.RouteResponse], timeout)
```

Comment:
It would make the code a lot more complex than it needs to. Why do you think it could kill the node's performance? I understand why mixing state machines and await can be problematic but it doesn't apply here as the `CreateInvoiceActor` is not really a state machine, it has only one state and is basically just a `Future`.

Reply:
`Await.result` blocks a system thread, and this is a scarce resource. Simply never use it, except in tests when you really can't avoid it. See https://www.reactivesystems.eu/2019/02/19/dont-use-awaitresult.html for a high level explanation.

Reply:
The ScalaDoc even says that you shouldn't use it unless you can't avoid it: https://www.scala-lang.org/api/current/scala/concurrent/Await$.html